core/src/main/java/com/linecorp/armeria/server/http/auth/HttpAuthService.java

```diff
@@ -0,0 +1,78 @@
+package com.linecorp.armeria.server.http.auth;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.Objects;
+import java.util.function.Predicate;
+
+import com.linecorp.armeria.common.http.DefaultHttpResponse;
+import com.linecorp.armeria.common.http.HttpHeaders;
+import com.linecorp.armeria.common.http.HttpRequest;
+import com.linecorp.armeria.common.http.HttpResponse;
+import com.linecorp.armeria.common.http.HttpStatus;
+import com.linecorp.armeria.server.DecoratingService;
+import com.linecorp.armeria.server.Service;
+import com.linecorp.armeria.server.ServiceRequestContext;
+
+/**
+ * A {@link DecoratingService} that provides HTTP authorization functionality.
+ *
+ * @param <T> the type of the authorization token
+ * */
+public abstract class HttpAuthService<T>
+        extends DecoratingService<HttpRequest, HttpResponse, HttpRequest, HttpResponse> {
+
+    private final Predicate<? super T> predicate;
+
+    /**
+     * Creates a new instance that provides HTTP authorization functionality to {@code delegate}.
+     *
+     * @param predicate Credential validation criteria.
+     */
+    protected HttpAuthService(Service<? super HttpRequest, ? extends HttpResponse> delegate,
+                              Predicate<? super T> predicate) {
+        super(delegate);
+        requireNonNull(predicate, "predicate");
+        this.predicate = predicate;
+    }
+
+    public Predicate<? super T> predicate() {
+        return predicate;
+    }
+
+    /**
+     * Extracts Authorization Token instance from given {@code req}. If {@code req} is invalid, returns null.
+     * */
+    protected abstract T extractAuthToken(HttpHeaders headers);
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+
+        if (o == null || !(o instanceof HttpAuthService)) {
+            return false;
+        }
+
+        HttpAuthService httpAuthService = (HttpAuthService) o;
+        return Objects.equals(predicate, httpAuthService.predicate);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(predicate);
+    }
+
+    @Override
+    public HttpResponse serve(ServiceRequestContext ctx, HttpRequest req) throws Exception {
+        T authToken = extractAuthToken(req.headers());
+        if (authToken != null && predicate.test(authToken)) {
+            return delegate().serve(ctx, req);
+        } else {
+            final DefaultHttpResponse res = new DefaultHttpResponse();
+            res.respond(HttpStatus.UNAUTHORIZED);
```

Comment:
How about moving this to an overridable method?

It's rare, but e.g., basic auth should return a response that causes the browser to ask for a password.